test/corpus/functions.txt

```diff
@@ -391,3 +391,499 @@ WHERE (
+    (statement
+     (select
+      (keyword_select)
+      (select_expression
+       (term
+        (literal)))))))))
+
+================================================================================
+Precedence between string body and `create table` with string options
+================================================================================
+
+create or replace function public.fn()
+  returns int
+  language sql
+as 'create table x (id int) row_format=dynamic';
+
+--------------------------------------------------------------------------------
+
+(program
+ (statement
+  (create_function
+   (keyword_create)
+   (keyword_or)
+   (keyword_replace)
+   (keyword_function)
+   (identifier)
+   (identifier)
+   (keyword_returns)
+   (int (keyword_int))
+   (function_language (keyword_sql))
+   (function_body
+    (keyword_as)
+    (statement
+     (create_table
+      (keyword_create)
+      (keyword_table)
+      (table_reference
+       (identifier))
+      (column_definitions
+       (column_definition
+        (identifier)
+        (int
+         (keyword_int))))
+      (table_option (identifier) (identifier))))))))
+
+================================================================================
+With `begin atomic`
+================================================================================
+
+create or replace function public.fn()
+  returns int
+  language sql
+begin atomic
+  return 1;
+end;
+
+--------------------------------------------------------------------------------
+
+(program
+ (statement
+  (create_function
+   (keyword_create)
+   (keyword_or)
+   (keyword_replace)
+   (keyword_function)
+   (identifier)
+   (identifier)
+   (keyword_returns)
+   (int (keyword_int))
+   (function_language (keyword_sql))
+   (function_body
+    (keyword_begin)
+    (keyword_atomic)
+    (keyword_return)
+    (literal)
+    (keyword_end)))))
+
+================================================================================
+Dollar quoting a simple PLPGSQL body
+================================================================================
+
+create or replace function public.fn()
+  returns int
+  language plpgsql
+as $function$
+begin
+  return 1;
+end;
+$function$;
+
+--------------------------------------------------------------------------------
+
+(program
+ (statement
+  (create_function
+   (keyword_create)
+   (keyword_or)
+   (keyword_replace)
+   (keyword_function)
+   (identifier)
+   (identifier)
+   (keyword_returns)
+   (int (keyword_int))
+   (function_language (keyword_plpgsql))
+   (function_body
+    (keyword_as)
+    (dollar_quote)
+    (keyword_begin)
+    (keyword_return)
+    (literal)
+    (keyword_end)
+    (dollar_quote)))))
+
+================================================================================
+Variable declarations
+================================================================================
+
+create or replace function public.fn()
+  returns int
+  language plpgsql
```

Comment:
I've never worked with functions in SQL before, or plpgsql. A few naive questions:

Would this make sense as a separate grammar that imports this one? Similar to how the `typescript` grammar imports the `javascript` grammar.

I'm happy to merge this into this parser but it feels like a different thing so I'm curious on if it's valuable to separate them to help manage complexity.

Reply:
`create function` is in the SQL standard, but I believe the function _body_ was originally defined as a string, until the more recent introduction of `begin atomic` for SQL-language (that is, without procedural extensions) functions.

An SQL grammar does need to support `create function` and the SQL-standard `return` etc for completeness' sake, but that's a good callout on the subject of superset grammars! The one caveat is that reserving stuff like `if`/`elsif`/`else` and `raise` for tree-sitter-plpgsql or a generic tree-sitter-procedural-sql would mean gaps in coverage, especially for the syntax highlighting use case.

Reply:
> The one caveat is that reserving stuff like if/elsif/else and raise for tree-sitter-plpgsql or a generic tree-sitter-procedural-sql would mean gaps in coverage, especially for the syntax highlighting use case.

We'd essentially need a to capture the `(function_body)` node use the separate parser to parse that node specifically. In neovim we'd use an `injection` query to do that.

For now, let's merge this in but if this part of the code grows complex we can investigate splitting it into a secondary parser.